### PR TITLE
Adding an example on handwriting recognition

### PR DESCRIPTION
**Pull request by sayakpaul**
A Colab Notebook is available [here](https://colab.research.google.com/github/sayakpaul/Handwriting-Recognizer-in-Keras/blob/pr/handwriting_recognition.ipynb).

Cc: @AakashKumarNain

### CODE REVIEW

File: examples/vision/handwriting_recognition.py [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+
+"""
+## Model
+
+Our model will use the CTC loss as an endpoint layer. For a detailed understanding of the
+CTC loss, refer to [this post](https://distill.pub/2017/ctc/). 
+"""
+
+
+class CTCLayer(keras.layers.Layer):
```
Comment (haifeng-jin) [inline examples/vision/handwriting_recognition.py:338]: Do we have to make the loss a `Layer` instead of a loss function or `Loss` subclass?

**Review by haifeng-jin — CHANGES_REQUESTED** [review]
Thank you for your contribution!

File: examples/vision/handwriting_recognition.py [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+Title: Handwriting Recognition
+Authors: [A_K_Nain](https://twitter.com/A_K_Nain), [Sayak Paul](https://twitter.com/RisingSayak)
+Date created: 2021/08/16
+Last modified: 2021/08/16
+Description: Training a handwriting recognition model with variable-length sequences.
+"""
+"""
+## Introduction
+
+This example shows how the [Captcha OCR](https://keras.io/examples/vision/captcha_ocr/)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/handwriting_recognition.py:11]: I don't think this is required since it's a sequel example.

File: examples/vision/handwriting_recognition.py (hunk 1 above)
Comment (sayakpaul) [reply examples/vision/handwriting_recognition.py:338]: Our idea with this example is to keep it as close to the Captcha OCR example as possible while showing the bits that need to be changed. IMO, that helps to ensure a good reading experience. Ccing @AakashKumarNain if he has other points of view.

File: examples/vision/handwriting_recognition.py [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+example can be extended to the
+[IAM Dataset](https://fki.tic.heia-fr.ch/databases/iam-handwriting-database)
+that has variable length ground-truths. IAM Dataset is widely used across many OCR
+benchmarks so we hope this example serves as a good starting point. 
+"""
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/handwriting_recognition.py:18]: Done.

File: examples/vision/handwriting_recognition.py (hunk 1 above)
Comment (AakashKumarNain) [reply examples/vision/handwriting_recognition.py:340]: When defining a normal loss function, we generally assume that the shape of `y_true` and `y_pred` are same which isn't the case here. Defining it as en endpoint layer makes it easy to calculate the length of inputs and the labels on the fly during training, and then pass the same to the `ctc_batch_cost(...)`

**Review by haifeng-jin — APPROVED** [review]
Thank you for the updates! LGTM.

**Comment by sayakpaul** [comment]
@haifeng-jin,
@AakashKumarNain and I may have another component to add to this tutorial. We are working on that. So, please expect a bit of delay as we finalize things.

**Comment by haifeng-jin** [comment]
@sayakpaul Sure, just ping me when it's ready. Thank you!

**Comment by sayakpaul** [comment]
@haifeng-jin now, it's good to go for another round of review. We incorporated Edit Distance as an evaluation metric.

File: examples/vision/handwriting_recognition.py [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+    callbacks=[early_stopping_callback, edit_distance_callback],
+)
+
+
+"""
+## Inference
+"""
+
+# Get the prediction model by extracting layers till the output layer.
+prediction_model = keras.models.Model(
```
Comment (AakashKumarNain) [inline examples/vision/handwriting_recognition.py:520]: @sayakpaul redundant. We have already defined prediction model above

Comment (sayakpaul) [reply examples/vision/handwriting_recognition.py]: Took care of it. Thanks!

**Review by haifeng-jin — APPROVED** [review]
LGTM, Thanks for the update!

**Review by haifeng-jin — APPROVED** [review]

**Review by fchollet — APPROVED** [review]
Awesome example! Very strong follow-up to the original OCR example.

I added a few minor copyedits. Please add the generated files. Thanks @haifeng-jin for the review.

**Comment by sayakpaul** [comment]
@fchollet added the generated files. The only major change is the reduced number of epochs because my system was getting stalled after 10 epochs. Tried on a commodity GCP Notebook instance too but didn't help much. 

Cc: @AakashKumarNain

**Comment by AakashKumarNain** [comment]
Thanks for the review @haifeng-jin @fchollet . 

@sayakpaul should we run this on a bigger VM (only if it is required)?

**Comment by sayakpaul** [comment]
I don't think that's required. However, if you want to go right ahead.

**Comment by fchollet** [comment]
Yes, I think the restriction should be fine. Thank you!

**Comment by fchollet** [comment]
@sayakpaul @AakashKumarNain there were only 2 image files included in the PR, but the example is intended to have 3 figures (it's missing the last one). Please add the missing figure (in a new PR)

**Comment by sayakpaul** [comment]
> but the example is intended to have 3 figures (it's missing the last one). Please add the missing figure (in a new PR)

@fchollet it actually plots two figures and uses markdown for the other one. See [here](https://github.com/keras-team/keras-io/blob/master/examples/vision/handwriting_recognition.py#L245).
